### PR TITLE
Making HV compile and run on JDK 9

### PR DESCRIPTION
**Pull request by gunnarmorling**
Supersedes #457

### CODE REVIEW

**Comment by gunnarmorling** [comment]
The change from #457 plus some adjustments to compile HV on JDK 9.

We discussed having a separate job for TCK-testing a JDK-8-built HV on JDK 9, and while that'd still be nice to have, it's actually much simpler if we are able to compile on 9, too. Unlike the original PR from some weeks ago, this basically specifies some type vars explicitly but does not alter any method signatures, so it appears acceptable to me.

**Comment by hferentschik** [comment]
> We discussed having a separate job for TCK-testing a JDK-8-built HV on JDK 9,

I am actually working on this. This be up shortly.

> it's actually much simpler if we are able to compile on 9, too.

Sure

> Unlike the original PR from some weeks ago, this basically specifies some type vars explicitly but does not alter any method signatures, so it appears acceptable to me.

+1 That's better

**Comment by gunnarmorling** [comment]
Pushed an update, @hferentschik:

* HV-1049: Ignoring annotations from "jdk.internal"; Thus avoiding the need to the policy file (removed that from @ctomc's commit)
* HV-1048: Obtaining the Java version; added some tests as suggested
* HV-1034: Make HV compile on JDK 9

File: engine/src/main/java/org/hibernate/validator/internal/util/Version.java [hunk 1]
```diff
@@ -30,10 +43,21 @@ public static void touch() {
...
+
+			if ( matcher.find() ) {
+				return Integer.valueOf( matcher.group( 1 ) );
+			}
+		}
 
-		return Integer.parseInt( specificationVersion[1] );
+		// Cannot determine Java version; Assuming 1.6, not enabling the 1.8-only features
+		LOG.unknownJvmVersion( specificationVersion );
+		return 6;
```
Comment (hferentschik) [inline engine/src/main/java/org/hibernate/validator/internal/util/Version.java]: interesting, defaulting to 6 if there version string is unexpected. I guess that is reasonable.

**Comment by hferentschik** [comment]
So is this pull request supposed to make the whole build work or just the compilation of the engine module? When I am trying to run the full build I get:

    Failed tests: 
      ConstraintValidationProcessorTest.hibernateValidatorProvidedCustomConstraints:110     Sets differ: expected [DiagnosticExpectation [kind=ERROR, lineNumber=55],     DiagnosticExpectation [kind=ERROR, lineNumber=56], DiagnosticExpectation [    kind=ERROR, lineNumber=57], DiagnosticExpectation [kind=ERROR, lineNumber=58],     DiagnosticExpectation [kind=ERROR, lineNumber=63], DiagnosticExpectation [    kind=ERROR, lineNumber=64], DiagnosticExpectation [kind=ERROR, lineNumber=65],     DiagnosticExpectation [kind=ERROR, lineNumber=66], DiagnosticExpectation [    kind=ERROR, lineNumber=59], DiagnosticExpectation [kind=ERROR, lineNumber=60],     DiagnosticExpectation [kind=ERROR, lineNumber=61], DiagnosticExpectation [    kind=ERROR, lineNumber=62], DiagnosticExpectation [kind=ERROR, lineNumber=67],     DiagnosticExpectation [kind=ERROR, lineNumber=68], DiagnosticExpectation [    kind=ERROR, lineNumber=69]] but got [DiagnosticExpectation [kind=ERROR,     lineNumber=55], DiagnosticExpectation [kind=ERROR, lineNumber=56],     DiagnosticExpectation [kind=ERROR, lineNumber=57], DiagnosticExpectation [    kind=ERROR, lineNumber=58], DiagnosticExpectation [kind=ERROR, lineNumber=63],     DiagnosticExpectation [kind=ERROR, lineNumber=64], DiagnosticExpectation [    kind=NOTE, lineNumber=-1], DiagnosticExpectation [kind=ERROR, lineNumber=65],     DiagnosticExpectation [kind=ERROR, lineNumber=66], DiagnosticExpectation [    kind=ERROR, lineNumber=59], DiagnosticExpectation [kind=ERROR, lineNumber=60],     DiagnosticExpectation [kind=ERROR, lineNumber=61], DiagnosticExpectation [    kind=ERROR, lineNumber=62], DiagnosticExpectation [kind=ERROR, lineNumber=67],     DiagnosticExpectation [kind=ERROR, lineNumber=68], DiagnosticExpectation [    kind=ERROR, lineNumber=69]]
      ConstraintValidationProcessorTest.timeConstraintsAllowedAtJodaTypes:531 Sets     differ: expected [DiagnosticExpectation [kind=ERROR, lineNumber=51],     DiagnosticExpectation [kind=ERROR, lineNumber=50]] but got [DiagnosticExpectation [    kind=ERROR, lineNumber=51], DiagnosticExpectation [kind=NOTE, lineNumber=-1],     DiagnosticExpectation [kind=ERROR, lineNumber=50]]
    
    Tests run: 34, Failures: 2, Errors: 0, Skipped: 0

**Comment by gunnarmorling** [comment]
> So is this pull request supposed to make the whole build work

Yes, full build actually. Seems you get an extra diagnostic of type "NOTE" in these two processors. It's a bit fugly to actually see those, probably easiest in the debugger. I had the same issue originally and it was about usage of deprecated APIs (reflected by a compiler warning). I meant to disable these via `@SuppressWarnings("deprecation") and indeed this works for me and on CI. So it'd be interesting whether a) the deprecation warning suppression doesn't work for your or b) you see another kind of diagnostic there.

**Comment by hferentschik** [comment]
Hmm, seems indeed there is an additional diagnostic message:

    timeConstraintsAllowedAtJodaTypes(org.hibernate.validator.ap.    ConstraintValidationProcessorTest)  Time elapsed: 0.02 sec  <<< FAILURE!
    java.lang.AssertionError: Sets differ: expected [DiagnosticExpectation [kind=ERROR,     lineNumber=51], DiagnosticExpectation [kind=ERROR, lineNumber=50]] but got [    DiagnosticExpectation [kind=ERROR, lineNumber=51], DiagnosticExpectation [kind=NOTE,     lineNumber=-1], DiagnosticExpectation [kind=ERROR, lineNumber=50]]
    	at org.testng.Assert.fail(Assert.java:94)
    	at org.testng.Assert.assertEquals(Assert.java:710)
    	at org.testng.Assert.assertEquals(Assert.java:688)
    	at org.hibernate.validator.ap.testutil.CompilerTestHelper.    assertThatDiagnosticsMatch(CompilerTestHelper.java:227)
    	at org.hibernate.validator.ap.ConstraintValidationProcessorTest.    timeConstraintsAllowedAtJodaTypes(ConstraintValidationProcessorTest.java:531)    

It seems with Java 8 I get the full build to work (hence also CI succeeding). With Java 9, I get these test failures. Are you sure you did your full build with Java 9 and not 8?

**Comment by gunnarmorling** [comment]
Ah, of course CI is on Java 8 :/ But I'm positive it's passing locally on
JDK 9 for me. Can you examine the diagnostics in the debugger?

**Comment by hferentschik** [comment]
> But I'm positive it's passing locally on JDK 9 for me. Can you examine the diagnostics in the debugger?

Ok. I am looking into it.

File: engine/src/main/java/org/hibernate/validator/internal/util/Version.java (hunk 1 above)
Comment (ctomc) [inline engine/src/main/java/org/hibernate/validator/internal/util/Version.java]: We use similar stuff in WildFly, but in master defaulting to 8.

File: engine/src/main/java/org/hibernate/validator/internal/metadata/provider/AnnotationMetaDataProvider.java [hunk 2]
```diff
@@ -564,6 +565,13 @@ else if ( parameterAnnotation.annotationType().equals( UnwrapValidatedValue.clas
 	protected <A extends Annotation> List<ConstraintDescriptorImpl<?>> findConstraintAnnotations(Member member,
 			A annotation,
 			ElementType type) {
+		// HV-1049 Methods of java.lang.Object (and potentially other JDK classes) are annotated with annotations such
+		// as jdk.internal.HotSpotIntrinsicCandidate on JDK 9; They cannot be constraint annotation so skip them right
+		// here, as for the proper check we'd need package access permission for "jdk.internal"
+		if ( annotation.annotationType().getPackage().getName().equals( "jdk.internal" ) ) {
```
Comment (ctomc) [inline engine/src/main/java/org/hibernate/validator/internal/metadata/provider/AnnotationMetaDataProvider.java]: Interesting, so this was the cause for the SM failures. good catch!

**Comment by hferentschik** [comment]
Ok, I merged the commits for HV-1048 and HV-1049. I am still seeing issues with the annotation processor and JDK 9. I will take the commit for HV-1034 and create a separate pull request (after figuring out what's going on)

**Pull request closed without merging**